<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 3.2//EN">
<HTML>
 <HEAD>
   <TITLE> [mod_python] Session Pickling Error II - 3.2.2b
   </TITLE>
   <LINK REL="Index" HREF="index.html" >
   <LINK REL="made" HREF="mailto:mod_python%40modpython.org?Subject=%5Bmod_python%5D%20Session%20Pickling%20Error%20II%20-%203.2.2b&In-Reply-To=c298f2d70510032243p222faee1t%40mail.gmail.com">
   <META NAME="robots" CONTENT="index,nofollow">
   <META http-equiv="Content-Type" content="text/html; charset=us-ascii">
   <LINK REL="Previous"  HREF="019160.html">
   <LINK REL="Next"  HREF="019165.html">
 </HEAD>
 <BODY BGCOLOR="#ffffff">
   <H1>[mod_python] Session Pickling Error II - 3.2.2b</H1>
<table border=0 width="100%"><tr><td valign="top">
    <B>GATOR AIDE</B> 
    <A HREF="mailto:mod_python%40modpython.org?Subject=%5Bmod_python%5D%20Session%20Pickling%20Error%20II%20-%203.2.2b&In-Reply-To=c298f2d70510032243p222faee1t%40mail.gmail.com"
       TITLE="[mod_python] Session Pickling Error II - 3.2.2b">gator_aide2000 at yahoo.com
       </A><BR>
    <I>Tue Oct  4 02:17:56 EDT 2005</I>
    <P><UL>
        <LI>Previous message: <A HREF="019160.html">[mod_python] Session Pickling Error II - 3.2.2b
</A></li>
        <LI>Next message: <A HREF="019165.html">[mod_python] Session Pickling Error II - 3.2.2b
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#19163">[ date ]</a>
              <a href="thread.html#19163">[ thread ]</a>
              <a href="subject.html#19163">[ subject ]</a>
              <a href="author.html#19163">[ author ]</a>
         </LI>
       </UL>
    <HR>  
<!--beginarticle-->
<PRE>Actually, the same exact code works using mod_python
version 3.1.4. 

I only get get the error when I run the code using the
latest version of mod_python (3.2.2b).

I guess I can work around the problem by storing
strings or something, but it seems odd that I only get
the error with the newest version. 

Sorry if that wasn't clear in the original post.

Thanks for all of the feedback.

Mike

--- Nicolas Lehuen &lt;<A HREF="http://mailman.modpython.org/mailman/listinfo/mod_python">nicolas.lehuen at gmail.com</A>&gt; wrote:

&gt;<i> Graham,
</I>&gt;<i> 
</I>&gt;<i> Mr Gator Aide had the same problem with mod_python
</I>&gt;<i> 3.1.4, though the message
</I>&gt;<i> was a bit different. The point is that he is trying
</I>&gt;<i> to store too much in the
</I>&gt;<i> session and that he should work around that. The new
</I>&gt;<i> import mechanism may
</I>&gt;<i> make things a little more difficult, but the core
</I>&gt;<i> problem is in the way he
</I>&gt;<i> uses sessions, not in mod_python.
</I>&gt;<i> 
</I>&gt;<i> Regards,
</I>&gt;<i> Nicolas
</I>&gt;<i> 
</I>&gt;<i> 2005/10/4, Graham Dumpleton &lt;<A HREF="http://mailman.modpython.org/mailman/listinfo/mod_python">grahamd at dscpl.com.au</A>&gt;:
</I>&gt;<i> &gt;
</I>&gt;<i> &gt; GATOR AIDE wrote ..
</I>&gt;<i> &gt; &gt; After updating mod_python 3.1.4 to 3.2.2b I am
</I>&gt;<i> getting
</I>&gt;<i> &gt; &gt; an error:
</I>&gt;<i> &gt; &gt;
</I>&gt;<i> &gt; &gt; PythonHandler mod_python.publisher: TypeError:
</I>&gt;<i> can't
</I>&gt;<i> &gt; &gt; pickle function objects
</I>&gt;<i> &gt; &gt;
</I>&gt;<i> &gt; &gt; I am storing a function in the session and
</I>&gt;<i> calling
</I>&gt;<i> &gt; &gt; session.save(). (Which is when the error is
</I>&gt;<i> raised.)
</I>&gt;<i> &gt; &gt;
</I>&gt;<i> &gt; &gt; Using mod_python 3.1.4 the code works just fine
</I>&gt;<i> but
</I>&gt;<i> &gt; &gt; after the update I am now seeing this error.
</I>&gt;<i> &gt; &gt;
</I>&gt;<i> &gt; &gt; The code simply stores a dictionary and few
</I>&gt;<i> function
</I>&gt;<i> &gt; &gt; callbacks in the session and loops through
</I>&gt;<i> several
</I>&gt;<i> &gt; &gt; functions until it is done processing.
</I>&gt;<i> &gt; &gt;
</I>&gt;<i> &gt; &gt; Not sure if anyone else has had this problem? I
</I>&gt;<i> &gt; &gt; running centos 2.4.21-27.0.4.ELsmp on a x86_64
</I>&gt;<i> &gt; &gt; platform. Python 2.4.2 (same error w/2.4.1).
</I>&gt;<i> Apache
</I>&gt;<i> &gt; &gt; 2.0.46.
</I>&gt;<i> &gt; &gt;
</I>&gt;<i> &gt; &gt; I wasn't sure if this is the right place to post
</I>&gt;<i> this,
</I>&gt;<i> &gt; &gt; so I apologize if I should have posted the
</I>&gt;<i> question
</I>&gt;<i> &gt; &gt; elsewhere.
</I>&gt;<i> &gt;
</I>&gt;<i> &gt; I am going to hazard a guess and say that pickle
</I>&gt;<i> probably only allows
</I>&gt;<i> &gt; you to pickle a function objects where the
</I>&gt;<i> function object resides in a
</I>&gt;<i> &gt; module which was imported using &quot;import&quot;, ie.,
</I>&gt;<i> lives in &quot; sys.modules&quot;.
</I>&gt;<i> &gt;
</I>&gt;<i> &gt; In mod_python 3.2, mod_python.publisher handler
</I>&gt;<i> modules are no longer
</I>&gt;<i> &gt; imported using the Python &quot;import&quot; mechanism but
</I>&gt;<i> are instead read in as
</I>&gt;<i> &gt; strings and executed against a dictionary of an
</I>&gt;<i> empty module which does
</I>&gt;<i> &gt; not actually exist in sys.modules.
</I>&gt;<i> &gt;
</I>&gt;<i> &gt; def build(self, key, name, opened, entry):
</I>&gt;<i> &gt; try:
</I>&gt;<i> &gt; module = new.module(re_not_word.sub('_',key))
</I>&gt;<i> &gt; module.__file__ = key
</I>&gt;<i> &gt; exec opened in module.__dict__
</I>&gt;<i> &gt; return module
</I>&gt;<i> &gt; finally:
</I>&gt;<i> &gt; opened.close()
</I>&gt;<i> &gt;
</I>&gt;<i> &gt; The error would seem reasonable under these
</I>&gt;<i> circumstances because if it
</I>&gt;<i> &gt; were pickled, it wouldn't perhaps be unable to
</I>&gt;<i> unpickle it later as it
</I>&gt;<i> &gt; wouldn't know the name of a module in
</I>&gt;<i> &quot;sys.modules&quot; to associate the
</I>&gt;<i> &gt; function reference to.
</I>&gt;<i> &gt;
</I>&gt;<i> &gt; Are you sure you can't use a string reference or
</I>&gt;<i> something else as a
</I>&gt;<i> &gt; handle
</I>&gt;<i> &gt; to refer to the callback function rather than an
</I>&gt;<i> actual function object?
</I>&gt;<i> &gt;
</I>&gt;<i> &gt; Graham
</I>&gt;<i> &gt; _______________________________________________
</I>&gt;<i> &gt; Mod_python mailing list
</I>&gt;<i> &gt; <A HREF="http://mailman.modpython.org/mailman/listinfo/mod_python">Mod_python at modpython.org</A>
</I>&gt;<i> &gt;
</I>&gt;<i>
</I><A HREF="http://mailman.modpython.org/mailman/listinfo/mod_python">http://mailman.modpython.org/mailman/listinfo/mod_python</A>
&gt;<i> &gt;
</I>&gt;<i> 
</I>


		
__________________________________ 
Yahoo! Mail - PC Magazine Editors' Choice 2005 
<A HREF="http://mail.yahoo.com">http://mail.yahoo.com</A>
</PRE>






<!--endarticle-->
    <HR>
    <P><UL>
        <!--threads-->
	<LI>Previous message: <A HREF="019160.html">[mod_python] Session Pickling Error II - 3.2.2b
</A></li>
	<LI>Next message: <A HREF="019165.html">[mod_python] Session Pickling Error II - 3.2.2b
</A></li>
         <LI> <B>Messages sorted by:</B> 
              <a href="date.html#19163">[ date ]</a>
              <a href="thread.html#19163">[ thread ]</a>
              <a href="subject.html#19163">[ subject ]</a>
              <a href="author.html#19163">[ author ]</a>
         </LI>
       </UL>

</td>
<td align="right" valign="top">
<script type="text/javascript"><!--
google_ad_client = "pub-9718360309690383";
google_ad_width = 120;
google_ad_height = 600;
google_ad_format = "120x600_as";
google_color_border = "CCCCCC";
google_color_bg = "FFFFFF";
google_color_link = "000000";
google_color_url = "666666";
google_color_text = "333333";
//--></script>
<script type="text/javascript"
  src="DISABLEhttp://pagead2.googlesyndication.com/pagead/show_ads.js">
</script>
</td>
</table>

<hr>
<a href="http://mailman.modpython.org/mailman/listinfo/mod_python">More information about the Mod_python
mailing list</a><br>
</body></html>
